src/storage.rs

```diff
@@ -2256,67 +2254,74 @@ impl Cache {
                 // Check if there are any relevant versions which are not present in
                 // the local cache. If none are missing, we have everything cached
                 // and can continue as normal.
-                let missing_versions: Vec<_> = versions
-                    .iter()
-                    .filter(|&&v| !published.versions.iter().any(|p| &p.num == v))
-                    .collect();
-                if missing_versions.is_empty() {
+                let missing_versions: Option<Vec<_>> = versions.map(|v| {
+                    v.into_iter()
+                        .filter(|&v| !published.versions.iter().any(|p| &p.num == v))
+                        .collect()
+                });
+                // If versions were specified and there were no missing versions, return
+                // immediately.
+                if missing_versions
+                    .as_ref()
+                    .map(|v| v.is_empty())
+                    .unwrap_or(false)
+                {
                     info!("using cached publisher info for {name} - relevant versions in cache");
-                    return Ok(published.versions.clone());
+                    return Ok(guard);
                 }
 
                 // If we last fetched this package's published versions less than a
                 // day ago, double-check if the versions we care about appear in the
                 // local crates.io index.
-                // If none of the versions appear in the local index, we can skip
-                // fetching. This should help in cases where a crate is marked as
-                // `audit-as-crates-io` but is not actually published, as we'll
-                // never find publisher information in those cases.
+                // If none of the versions appear in the local index or no versions were specified,
+                // we can skip fetching. This should help in cases where a crate is marked as
+                // `audit-as-crates-io` but is not actually published, as we'll never find
+                // publisher information in those cases.
                 if now - published.last_fetched
                     < chrono::Duration::days(NONINDEX_VERSION_PUBLISHER_REFRESH_DAYS)
                 {
                     if let Some(index_crate) = self.query_package_from_index(name) {
                         if missing_versions
-                            .iter()
-                            .all(|version| exact_version(&index_crate, version).is_none())
+                            .map(|v| {
+                                v.iter()
+                                    .all(|version| exact_version(&index_crate, version).is_none())
+                            })
+                            .unwrap_or(true)
                         {
                             info!("using cached publisher info for {name} - missing versions appear unpublished");
```

Comment:
Might be nice to change this `info!` log in the case where no versions are specified.